21.4.1 fails to build with ffmpeg >= 5.0

21.4.1 while it builds fine with ffmpeg 4.4.1 it fails to build with the just released ffmpeg 5.0:

```
[...]
[ 51%] Building CXX object code/CMakeFiles/code.dir/cutscene/ffmpeg/VideoDecoder.cpp.o
cd /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/code && /usr/bin/x86_64-pc-linux-gnu-c++ -DANTLR4CPP_STATIC -DCMAKE_PCH -DNDEBUG -DROCKET_STATIC_LIB -DSCP_UNIX -DUSE_OPENAL -DVK_NO_PROTOTYPES -DVULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1 -DWITH_FFMPEG -DWITH_OPENGL -DWITH_VULKAN -D_REENTRANT -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS=1 -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/freespace2 -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/generated_source -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/generated_source/code -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/code -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/code/embedded_files -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/discord/src/../include -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/libRocket/Include -I/usr/x86_64-pc-linux-gnu/include/freetype2 -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/parsers -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/antlr4-cpp-runtime/runtime/src -I/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/mongoose -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/opengl/gl/glad/include -isystem /usr/x86_64-pc-linux-gnu/include/AL -isystem /usr/x86_64-pc-linux-gnu/include/lua5.1 -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/lz4 -isystem /usr/x86_64-pc-linux-gnu/include/SDL2 -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/utfcpp -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/md5 -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/libpcp/include -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/vulkan-hpp -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/mdns -isystem /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/lib/variant -march=native -O2 -pipe   -W -Wall -Wextra -Wno-maybe-uninitialized -funroll-loops -fsigned-char -Wno-unknown-pragmas -fdata-sections -ffunction-sections  -Wformat-security -Wreturn-type -Wno-deprecated -Wno-char-subscripts -Wno-stringop-truncation -Wno-stringop-overflow -O2 -Wno-unused-variable -Wno-unused-but-set-variable -Wno-array-bounds -Wno-empty-body -Wno-clobbered  -Wno-unused-parameter -std=gnu++11 -Winvalid-pch -include /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build/code/CMakeFiles/code.dir/cmake_pch.hxx -MD -MT code/CMakeFiles/code.dir/cutscene/ffmpeg/VideoDecoder.cpp.o -MF CMakeFiles/code.dir/cutscene/ffmpeg/VideoDecoder.cpp.o.d -o CMakeFiles/code.dir/cutscene/ffmpeg/VideoDecoder.cpp.o -c /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp: In destructor 'cutscene::ffmpeg::DecoderStatus::~DecoderStatus()':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:15:17: error: 'avcodec_close' was not declared in this scope; did you mean 'avio_close'?
   15 |                 avcodec_close(videoCodecCtx);
      |                 ^~~~~~~~~~~~~
      |                 avio_close
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp: In function 'double {anonymous}::getFrameRate(AVStream*, AVCodecContext*)':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:59:44: error: invalid use of incomplete type 'struct AVCodecContext'
   59 |                 fps = 1.0 / av_q2d(codecCtx->time_base);
      |                                            ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegContext.h:8,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:6:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:195:8: note: forward declaration of 'struct AVCodecContext'
  195 | struct AVCodecContext;
      |        ^~~~~~~~~~~~~~
make[2]: *** [code/CMakeFiles/code.dir/build.make:1370: code/CMakeFiles/code.dir/cutscene/ffmpeg/AudioDecoder.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp: In function 'std::unique_ptr<cutscene::ffmpeg::DecoderStatus> cutscene::ffmpeg::{anonymous}::initializeStatus(std::unique_ptr<cutscene::ffmpeg::InputStream>&, std::unique_ptr<cutscene::ffmpeg::InputStream>&, const cutscene::PlaybackProperties&)':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:232:81: error: invalid conversion from 'AVCodec**' to 'const AVCodec**' [-fpermissive]
  232 |         auto videoStream = av_find_best_stream(ctx, AVMEDIA_TYPE_VIDEO, -1, -1, &status->videoCodec, 0);
      |                                                                                 ^~~~~~~~~~~~~~~~~~~
      |                                                                                 |
      |                                                                                 AVCodec**
In file included from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegContext.h:8,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:6:
/usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:2161:41: note:   initializing argument 5 of 'int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)'
 2161 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:248:93: error: invalid conversion from 'AVCodec**' to 'const AVCodec**' [-fpermissive]
  248 |                 audioStream = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, videoStream, &status->audioCodec, 0);
      |                                                                                             ^~~~~~~~~~~~~~~~~~~
      |                                                                                             |
      |                                                                                             AVCodec**
In file included from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegContext.h:8,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:6:
/usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:2161:41: note:   initializing argument 5 of 'int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)'
 2161 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:324:33: error: 'avcodec_alloc_context3' was not declared in this scope; did you mean 'avio_alloc_context'?
  324 |         status->videoCodecCtx = avcodec_alloc_context3(status->videoCodec);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~
      |                                 avio_alloc_context
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:17:17: error: 'avcodec_free_context' was not declared in this scope; did you mean 'avformat_free_context'?
   17 |                 avcodec_free_context(&videoCodecCtx);
      |                 ^~~~~~~~~~~~~~~~~~~~
      |                 avformat_free_context
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:327:15: error: 'avcodec_parameters_to_context' was not declared in this scope; did you mean 'avcodec_parameters_copy'?
  327 |         err = avcodec_parameters_to_context(status->videoCodecCtx, status->videoStream->codecpar);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |               avcodec_parameters_copy
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:27:17: error: 'avcodec_close' was not declared in this scope; did you mean 'avio_close'?
   27 |                 avcodec_close(audioCodecCtx);
      |                 ^~~~~~~~~~~~~
      |                 avio_close
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:338:15: error: 'avcodec_open2' was not declared in this scope; did you mean 'avio_open2'?
  338 |         err = avcodec_open2(status->videoCodecCtx, status->videoCodec, nullptr);
      |               ^~~~~~~~~~~~~
      |               avio_open2
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:382:61: error: invalid conversion from 'const AVCodec*' to 'AVCodec*' [-fpermissive]
  382 |                 status->subtitleCodec = avcodec_find_decoder(status->subtitleCodecPars.codec_id);
      |                                         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                             |
      |                                                             const AVCodec*
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::SubtitleDecoder::decodePacket(AVPacket*)':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:24:20: error: aggregate 'AVSubtitle subtitle' has incomplete type and cannot be defined
   24 |         AVSubtitle subtitle;
      |                    ^~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:29:17: error: 'avcodec_free_context' was not declared in this scope; did you mean 'avformat_free_context'?
   29 |                 avcodec_free_context(&audioCodecCtx);
      |                 ^~~~~~~~~~~~~~~~~~~~
      |                 avformat_free_context
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:25:23: error: 'avcodec_decode_subtitle2' was not declared in this scope
   25 |         auto result = avcodec_decode_subtitle2(m_status->subtitleCodecCtx, &subtitle, &finishedFrame, packet);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:39:17: error: 'avcodec_close' was not declared in this scope; did you mean 'avio_close'?
   39 |                 avcodec_close(subtitleCodecCtx);
      |                 ^~~~~~~~~~~~~
      |                 avio_close
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::VideoDecoder::decodePacket(AVPacket*)':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp:108:31: error: 'avcodec_send_packet' was not declared in this scope; did you mean 'avcodec_get_name'?
  108 |                 send_result = avcodec_send_packet(m_status->videoCodecCtx, packet);
      |                               ^~~~~~~~~~~~~~~~~~~
      |                               avcodec_get_name
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:30:17: error: 'avsubtitle_free' was not declared in this scope
   30 |                 avsubtitle_free(&subtitle);
      |                 ^~~~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::SubtitleDecoder::finishDecoding()':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:40:20: error: aggregate 'AVSubtitle subtitle' has incomplete type and cannot be defined
   40 |         AVSubtitle subtitle;
      |                    ^~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.cpp:41:17: error: 'avcodec_free_context' was not declared in this scope; did you mean 'avformat_free_context'?
   41 |                 avcodec_free_context(&subtitleCodecCtx);
      |                 ^~~~~~~~~~~~~~~~~~~~
      |                 avformat_free_context
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp:110:23: error: 'avcodec_receive_frame' was not declared in this scope; did you mean 'avcodec_profile_name'?
  110 |                 while(avcodec_receive_frame(m_status->videoCodecCtx, m_decodeFrame) == 0) {
      |                       ^~~~~~~~~~~~~~~~~~~~~
      |                       avcodec_profile_name
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:43:28: error: 'avcodec_decode_subtitle2' was not declared in this scope
   43 |                 auto err = avcodec_decode_subtitle2(m_status->subtitleCodecCtx, &subtitle, &finishedFrame, &nullPacket);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::VideoDecoder::finishDecoding()':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp:129:9: error: 'avcodec_send_packet' was not declared in this scope; did you mean 'avcodec_get_name'?
  129 |         avcodec_send_packet(m_status->videoCodecCtx, nullptr);
      |         ^~~~~~~~~~~~~~~~~~~
      |         avcodec_get_name
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:51:17: error: 'avsubtitle_free' was not declared in this scope
   51 |                 avsubtitle_free(&subtitle);
      |                 ^~~~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp: In member function 'void cutscene::ffmpeg::SubtitleDecoder::pushSubtitleFrame(AVPacket*, AVSubtitle*)':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:55:21: error: invalid use of incomplete type 'struct AVSubtitle'
   55 |         if (subtitle->format != 1) {
      |                     ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:60:21: error: invalid use of incomplete type 'struct AVSubtitle'
   60 |         if (subtitle->num_rects < 1) {
      |                     ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:64:31: error: invalid use of incomplete type 'struct AVSubtitle'
   64 |         int64_t pts = subtitle->pts;
      |                               ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:74:50: error: invalid use of incomplete type 'struct AVSubtitle'
   74 |         auto start_time = packet_time + (subtitle->start_display_time / 1000.0);
      |                                                  ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:75:48: error: invalid use of incomplete type 'struct AVSubtitle'
   75 |         auto end_time = packet_time + (subtitle->end_display_time / 1000.0);
      |                                                ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:76:21: error: invalid use of incomplete type 'struct AVSubtitle'
   76 |         if (subtitle->end_display_time == 0 && m_status->subtitleStream->time_base.num != 0) {
      |                     ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:82:38: error: invalid use of incomplete type 'struct AVSubtitle'
   82 |         auto subtitle_rect = subtitle->rects[0];
      |                                      ^~
In file included from /usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:313,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:10,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/internal.h:5,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.h:3,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:4:
/usr/x86_64-pc-linux-gnu/include/libavcodec/codec.h:196:8: note: forward declaration of 'struct AVSubtitle'
  196 | struct AVSubtitle;
      |        ^~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp:133:28: error: 'avcodec_receive_frame' was not declared in this scope; did you mean 'avcodec_profile_name'?
  133 |                 auto ret = avcodec_receive_frame(m_status->videoCodecCtx, m_decodeFrame);
      |                            ^~~~~~~~~~~~~~~~~~~~~
      |                            avcodec_profile_name
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:83:36: error: 'SUBTITLE_BITMAP' was not declared in this scope
   83 |         if (subtitle_rect->type == SUBTITLE_BITMAP) {
      |                                    ^~~~~~~~~~~~~~~
make[2]: *** [code/CMakeFiles/code.dir/build.make:1402: code/CMakeFiles/code.dir/cutscene/ffmpeg/internal.cpp.o] Error 1
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:87:43: error: 'SUBTITLE_TEXT' was not declared in this scope
   87 |         } else if (subtitle_rect->type == SUBTITLE_TEXT) {
      |                                           ^~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::VideoDecoder::flushBuffers()':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/VideoDecoder.cpp:162:9: error: 'avcodec_flush_buffers' was not declared in this scope
  162 |         avcodec_flush_buffers(m_status->videoCodecCtx);
      |         ^~~~~~~~~~~~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:90:43: error: 'SUBTITLE_ASS' was not declared in this scope
   90 |         } else if (subtitle_rect->type == SUBTITLE_ASS) {
      |                                           ^~~~~~~~~~~~
make[2]: *** [code/CMakeFiles/code.dir/build.make:1386: code/CMakeFiles/code.dir/cutscene/ffmpeg/FFMPEGDecoder.cpp.o] Error 1
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp: In member function 'virtual void cutscene::ffmpeg::SubtitleDecoder::flushBuffers()':
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/SubtitleDecoder.cpp:133:9: error: 'avcodec_flush_buffers' was not declared in this scope
  133 |         avcodec_flush_buffers(m_status->subtitleCodecCtx);
      |         ^~~~~~~~~~~~~~~~~~~~~
make[2]: *** [code/CMakeFiles/code.dir/build.make:1434: code/CMakeFiles/code.dir/cutscene/ffmpeg/VideoDecoder.cpp.o] Error 1
make[2]: *** [code/CMakeFiles/code.dir/build.make:1418: code/CMakeFiles/code.dir/cutscene/ffmpeg/SubtitleDecoder.cpp.o] Error 1
make[2]: Leaving directory '/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build'
make[1]: Leaving directory '/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/build'
make[1]: *** [CMakeFiles/Makefile2:926: code/CMakeFiles/code.dir/all] Error 2
make: *** [Makefile:159: all] Error 2
```
Complete build log: [fs2_open-21.4.1-build.log](https://github.com/scp-fs2open/fs2open.github.com/files/7875564/fs2_open-21.4.1-build.log)

https://github.com/scp-fs2open/fs2open.github.com/pull/3927 partly fixes it and leaves us with `FFMPEGDecoder.cpp`

```
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:232:81: error: invalid conversion from 'AVCodec**' to 'const AVCodec**' [-fpermissive]
  232 |         auto videoStream = av_find_best_stream(ctx, AVMEDIA_TYPE_VIDEO, -1, -1, &status->videoCodec, 0);
      |                                                                                 ^~~~~~~~~~~~~~~~~~~
      |                                                                                 |
      |                                                                                 AVCodec**
In file included from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:11,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegContext.h:8,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:6:
/usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:2161:41: note:   initializing argument 5 of 'int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)'
 2161 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:248:93: error: invalid conversion from 'AVCodec**' to 'const AVCodec**' [-fpermissive]
  248 |                 audioStream = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, videoStream, &status->audioCodec, 0);
      |                                                                                             ^~~~~~~~~~~~~~~~~~~
      |                                                                                             |
      |                                                                                             AVCodec**
In file included from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegHeaders.h:11,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/libs/ffmpeg/FFmpegContext.h:8,
                 from /var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:6:
/usr/x86_64-pc-linux-gnu/include/libavformat/avformat.h:2161:41: note:   initializing argument 5 of 'int av_find_best_stream(AVFormatContext*, AVMediaType, int, int, const AVCodec**, int)'
 2161 |                         const AVCodec **decoder_ret,
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/var/tmp/paludis/build/games-simulation-fs2_open-21.4.1/work/fs2_open-21.4.1/code/cutscene/ffmpeg/FFMPEGDecoder.cpp:382:61: error: invalid conversion from 'const AVCodec*' to 'AVCodec*' [-fpermissive]
  382 |                 status->subtitleCodec = avcodec_find_decoder(status->subtitleCodecPars.codec_id);
      |                                         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                             |
      |                                                             const AVCodec*
```
which is due to the upstream change https://github.com/FFmpeg/FFmpeg/commit/46dac8cf3d250184ab4247809bc03f60e14f4c0c `av_find_best_stream now uses a const AVCodec ** parameter for the returned decoder. `

Since I'm not a coder I wasn't able to fix these.

--- a/code/cutscene/ffmpeg/FFMPEGDecoder.cpp
+++ b/code/cutscene/ffmpeg/FFMPEGDecoder.cpp
@@ -229,7 +229,7 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 
 	auto ctx = stream->m_ctx->ctx();
 
-	auto videoStream = av_find_best_stream(ctx, AVMEDIA_TYPE_VIDEO, -1, -1, &status->videoCodec, 0);
+	auto videoStream = av_find_best_stream(ctx, AVMEDIA_TYPE_VIDEO, -1, -1, nullptr, 0);
 	if (videoStream < 0) {
 		if (videoStream == AVERROR_STREAM_NOT_FOUND) {
 			mprintf(("FFmpeg: No video stream found in file!\n"));
@@ -242,10 +242,17 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 		return nullptr;
 	}
 
+	status->videoCodec = avcodec_find_decoder(ctx->streams[videoStream]->codecpar->codec_id);
+
+	if ( !status->videoCodec ) {
+		mprintf(("FFmpeg: Cannot find decoder for video stream!\n"));
+		return nullptr;
+	}
+
 	int audioStream = -1;
 
 	if (properties.with_audio) {
-		audioStream = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, videoStream, &status->audioCodec, 0);
+		audioStream = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, videoStream, nullptr, 0);
 		if (audioStream < 0) {
 			if (audioStream == AVERROR_STREAM_NOT_FOUND) {
 				mprintf(("FFmpeg: No audio stream found in file!\n"));
@@ -254,6 +261,12 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 			} else {
 				mprintf(("FFmpeg: Unknown error while finding audio stream!\n"));
 			}
+		}
+
+		status->audioCodec = avcodec_find_decoder(ctx->streams[audioStream]->codecpar->codec_id);
+
+		if ( !status->audioCodec ) {
+			mprintf(("FFmpeg: Cannot find decoder for audio stream!\n"));
 		}
 	}
 

--- a/code/cutscene/ffmpeg/FFMPEGDecoder.cpp
+++ b/code/cutscene/ffmpeg/FFMPEGDecoder.cpp
@@ -242,10 +242,19 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 		return nullptr;
 	}
 
-	status->videoCodec = avcodec_find_decoder(ctx->streams[videoStream]->codecpar->codec_id);
+	status->videoStreamIndex = videoStream;
+	status->videoStream = ctx->streams[videoStream];
+	status->videoCodecPars = getCodecParameters(status->videoStream);
+
+	status->videoCodec = avcodec_find_decoder(status->videoCodecPars.codec_id);
 
 	if ( !status->videoCodec ) {
 		mprintf(("FFmpeg: Cannot find decoder for video stream!\n"));
+		return nullptr;
+	}
+
+	if (properties.looping && status->videoCodecPars.codec_id == AV_CODEC_ID_INTERPLAY_VIDEO) {
+		mprintf(("FFmpeg: Looping is not supported for inteplay (MVE) movies!\n"));
 		return nullptr;
 	}
 
@@ -261,21 +270,19 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 			} else {
 				mprintf(("FFmpeg: Unknown error while finding audio stream!\n"));
 			}
+		} else {
+			status->audioStreamIndex = audioStream;
+			status->audioStream = ctx->streams[audioStream];
+			status->audioCodecPars = getCodecParameters(status->audioStream);
+
+			status->audioCodec = avcodec_find_decoder(status->audioCodecPars.codec_id);
+
+			if ( !status->audioCodec ) {
+				mprintf(("FFmpeg: Cannot find decoder for audio stream!\n"));
+				status->audioStreamIndex = -1;
+				audioStream = -1;
+			}
 		}
-
-		status->audioCodec = avcodec_find_decoder(ctx->streams[audioStream]->codecpar->codec_id);
-
-		if ( !status->audioCodec ) {
-			mprintf(("FFmpeg: Cannot find decoder for audio stream!\n"));
-		}
-	}
-
-	status->videoStreamIndex = videoStream;
-	status->videoStream = ctx->streams[videoStream];
-
-	if (audioStream >= 0) {
-		status->audioStreamIndex = audioStream;
-		status->audioStream = ctx->streams[audioStream];
 	}
 
 	if (subt) {
@@ -291,9 +298,13 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 				mprintf(("FFmpeg: Unknown error while finding subtitle stream!\n"));
 			}
 		} else {
+			status->subtitleStreamIndex = subtStream;
 			status->subtitleStream      = subtCtx->streams[subtStream];
 			status->externalSubtitles   = true;
-			status->subtitleStreamIndex = subtStream;
+			// Not really sure if this makes sense for subtitles but it shouldn't break anything...
+			status->subtitleCodecPars = getCodecParameters(status->subtitleStream);
+
+			status->subtitleCodec = avcodec_find_decoder(status->subtitleCodecPars.codec_id);
 		}
 	}
 
@@ -325,13 +336,6 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 		}
 	}
 
-    status->videoCodecPars = getCodecParameters(status->videoStream);
-
-	if (properties.looping && status->videoCodecPars.codec_id == AV_CODEC_ID_INTERPLAY_VIDEO) {
-		mprintf(("FFmpeg: Looping is not supported for inteplay (MVE) movies!\n"));
-		return nullptr;
-	}
-
 	int err;
 #if LIBAVCODEC_VERSION_INT > AV_VERSION_INT(57, 24, 255)
 	status->videoCodecCtx = avcodec_alloc_context3(status->videoCodec);
@@ -361,8 +365,6 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 
 	// Now initialize audio, if this fails it's not a fatal error
 	if (audioStream >= 0) {
-		status->audioCodecPars = getCodecParameters(status->audioStream);
-
 #if LIBAVCODEC_VERSION_INT > AV_VERSION_INT(57, 24, 255)
 		status->audioCodecCtx = avcodec_alloc_context3(status->audioCodec);
 
@@ -389,11 +391,6 @@ std::unique_ptr<DecoderStatus> initializeStatus(std::unique_ptr<InputStream>& st
 	}
 
 	if (status->subtitleStream != nullptr) {
-		// Not really sure if this makes sense for subtitles but it shouldn't break anything...
-		status->subtitleCodecPars = getCodecParameters(status->subtitleStream);
-
-		status->subtitleCodec = avcodec_find_decoder(status->subtitleCodecPars.codec_id);
-
 #if LIBAVCODEC_VERSION_INT > AV_VERSION_INT(57, 24, 255)
 		status->subtitleCodecCtx = avcodec_alloc_context3(status->subtitleCodec);
 

--- a/code/cutscene/ffmpeg/internal.h
+++ b/code/cutscene/ffmpeg/internal.h
@@ -38,7 +38,7 @@ struct DecoderStatus {
 	int subtitleStreamIndex = -1;
 	AVStream* subtitleStream = nullptr;
 	CodecContextParameters subtitleCodecPars;
-	AVCodec* subtitleCodec = nullptr;
+	const AVCodec* subtitleCodec = nullptr;
 	AVCodecContext* subtitleCodecCtx = nullptr;
 
 	DecoderStatus();

--- a/code/cutscene/ffmpeg/internal.h
+++ b/code/cutscene/ffmpeg/internal.h
@@ -24,13 +24,13 @@ struct DecoderStatus {
 	int videoStreamIndex = -1;
 	AVStream* videoStream = nullptr;
 	CodecContextParameters videoCodecPars;
-	AVCodec* videoCodec = nullptr;
+	const AVCodec* videoCodec = nullptr;
 	AVCodecContext* videoCodecCtx = nullptr;
 
 	int audioStreamIndex = -1;
 	AVStream* audioStream = nullptr;
 	CodecContextParameters audioCodecPars;
-	AVCodec* audioCodec = nullptr;
+	const AVCodec* audioCodec = nullptr;
 	AVCodecContext* audioCodecCtx = nullptr;
 
 	// Subtitles are a bit different since they max come from a different file

--- a/code/sound/ffmpeg/FFmpegWaveFile.cpp
+++ b/code/sound/ffmpeg/FFmpegWaveFile.cpp
@@ -132,18 +132,24 @@ FFmpegWaveFile::~FFmpegWaveFile()
 	m_ctx.reset();
 }
 
-AVCodec* sound::ffmpeg::FFmpegWaveFile::prepareOpened()
+const AVCodec* sound::ffmpeg::FFmpegWaveFile::prepareOpened()
 {
     using namespace libs::ffmpeg;
     auto ctx = m_ctx->ctx();
 
-    AVCodec* audio_codec = nullptr;
+    const AVCodec* audio_codec = nullptr;
 
-    m_audioStreamIndex   = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, -1, &audio_codec, 0);
+    m_audioStreamIndex   = av_find_best_stream(ctx, AVMEDIA_TYPE_AUDIO, -1, -1, nullptr, 0);
     if (m_audioStreamIndex < 0) {
         throw FFmpegException("Failed to find audio stream in file.");
     }
     m_audioStream = ctx->streams[m_audioStreamIndex];
+
+	audio_codec = avcodec_find_decoder(m_audioStream->codecpar->codec_id);
+
+	if ( !audio_codec ) {
+		throw FFmpegException("Failed to find decoder for audio stream in file.");
+	}
 
     int err;
 #if LIBAVCODEC_VERSION_INT > AV_VERSION_INT(57, 24, 255)
@@ -229,7 +235,7 @@ bool FFmpegWaveFile::Open(const char* pszFilename, bool keep_ext)
 
 		m_ctx    = FFmpegContext::createContext(cfp);
 
-        AVCodec* audio_codec = prepareOpened();
+        const AVCodec* audio_codec = prepareOpened();
 
 		nprintf(("SOUND", "SOUND => %s => Using codec %s (%s)\n", filename, audio_codec->long_name, audio_codec->name));
 	} catch (const FFmpegException& e) {
@@ -255,7 +261,7 @@ bool FFmpegWaveFile::OpenMem(const uint8_t* snddata, size_t snd_len)
 	try {
 		m_ctx    = FFmpegContext::createContextMem(snddata, snd_len);
 
-        AVCodec* audio_codec = prepareOpened();
+        const AVCodec* audio_codec = prepareOpened();
 
 		nprintf(("SOUND", "SOUND => %s => Using codec %s (%s)\n", "soundfile-in-memory", audio_codec->long_name, audio_codec->name));
 	} catch (const FFmpegException& e) {

--- a/code/sound/ffmpeg/FFmpegWaveFile.h
+++ b/code/sound/ffmpeg/FFmpegWaveFile.h
@@ -101,7 +101,7 @@ class FFmpegWaveFile : public IAudioFile {
 	void setResamplingProperties(const ResampleProperties& resampleProps) override;
 
   private:
-    AVCodec* prepareOpened();
+    const AVCodec* prepareOpened();
 
 	size_t handleDecodedFrame(AVFrame* av_frame, uint8_t* out_buffer, size_t buffer_size);
 

